Source/Core/DiscIO/FileSystemGCWii.cpp

```diff
@@ -19,83 +21,309 @@
 
 namespace DiscIO
 {
+// Set everything manually
+CFileInfoGCWii::CFileInfoGCWii(const u8* fst, bool wii, u32 fst_size, u32 index, u32 total_file_infos)
+	: m_fst(fst),
+	m_wii(wii),
+	m_fst_size(fst_size),
+	m_index(index),
+	m_total_file_infos(total_file_infos)
+{ }
+
+// For the root object only
+CFileInfoGCWii::CFileInfoGCWii(const u8* fst, bool wii, u32 fst_size)
+	: m_fst(fst),
+	m_wii(wii),
+	m_fst_size(fst_size),
+	m_index(0),
+	m_total_file_infos(GetSize())
+{ }
+
+// Copies data that is common to the whole file system
+CFileInfoGCWii::CFileInfoGCWii(const CFileInfoGCWii& file_info, u32 index)
+	: m_fst(file_info.m_fst),
+	m_wii(file_info.m_wii),
+	m_fst_size(file_info.m_fst_size),
+	m_index(FixIndex(index)),
+	m_total_file_infos(file_info.m_total_file_infos)
+{ }
+
+CFileInfoGCWii::~CFileInfoGCWii()
+{
+}
+
+
+u64 CFileInfoGCWii::GetFSTSize() const
+{
+	return (u64)m_fst_size << (m_wii ? 2 : 0);
+}
+
+u32 CFileInfoGCWii::FixIndex(u32 index) const
+{
+	if (GetFSTSize() / 0xC <= index)
+	{
+		// To avoid reading beyond the end of the FST later, the index is clamped if it's too large.
+		ERROR_LOG(DISCIO, "Index in file system is too large");
+		return (u32)(GetFSTSize() / 0xC - 1);
+	}
+
+	return index;
+}
+
+const u8* CFileInfoGCWii::GetAddress() const
+{
+	return m_fst + m_index * 0xC;
+}
+
+std::unique_ptr<IFileInfo> CFileInfoGCWii::Clone() const
+{
+	return std::unique_ptr<IFileInfo>(new CFileInfoGCWii(*this));
+}
+
+std::unique_ptr<IFileInfo> CFileInfoGCWii::Next() const
+{
+	return std::unique_ptr<IFileInfo>(new CFileInfoGCWii(*this, IsDirectory() ? GetSize() : m_index + 1));
+}
+
+IFileInfo::const_iterator CFileInfoGCWii::begin() const
+{
+	return const_iterator(std::unique_ptr<IFileInfo>(new CFileInfoGCWii(*this, m_index + 1)));
+}
+
+IFileInfo::const_iterator CFileInfoGCWii::end() const
+{
+	return const_iterator(Next());
+}
+
+// Reads the first u32 of this entry for Get(0), reads the u32 after that for Get(1), and so on.
+// Each entry contains 3 u32s (0xC bytes).
+u32 CFileInfoGCWii::Get(int n) const { return Common::swap32(*((u32*)(m_fst + m_index * 0xC) + n)); }
+u32 CFileInfoGCWii::GetNameOffset() const { return Get(0); }
+u32 CFileInfoGCWii::GetRawOffset() const { return Get(1); }
```

Comment:
Does it really matter for one-liners like that? IMO putting this on one line is more readable.